docker/cygnus-ngsi/Dockerfile

```diff
@@ -0,0 +1,111 @@
+ENV GIT_REV_CYGNUS develop
+
+ENV MVN_VER "3.2.5"
+ENV MVN_TGZ "apache-maven-${MVN_VER}-bin.tar.gz"
+ENV MVN_URL "http://www.eu.apache.org/dist/maven/maven-3/${MVN_VER}/binaries/${MVN_TGZ}"
+ENV MVN_HOME "/opt/apache-maven"
+
+ENV FLUME_VER "1.4.0"
+ENV FLUME_TGZ "apache-flume-${FLUME_VER}-bin.tar.gz"
+ENV FLUME_URL "http://archive.apache.org/dist/flume/${FLUME_VER}/${FLUME_TGZ}"
+ENV FLUME_HOME "/opt/apache-flume"
+
+ENV CONF_PATH "/opt/apache-flume/conf"
+ENV CONF_FILE "/opt/apache-flume/conf/agent.conf"
+ENV AGENT_NAME "cygnus-ngsi"
+ENV LOG_LEVEL "INFO"
+ENV LOG_APPENDER "console"
+
+ENV JAVA_HOME /usr/lib/jvm/java-1.6.0
+WORKDIR /opt
+
+RUN adduser --comment "${CYGNUS_USER}" ${CYGNUS_USER}
+
+# Install dependencies (Java, Maven, Flume)
+RUN yum -y install tar nc which git java-1.6.0-openjdk-devel && \
+    yum clean all && \
+    curl --remote-name --location --insecure --silent --show-error "${MVN_URL}" && \
+    tar xzvf ${MVN_TGZ} && \
+    mv apache-maven-${MVN_VER} ${MVN_HOME} && \
+    export MAVEN_OPTS="-Xmx512m -XX:MaxPermSize=128m" && \
+    curl --remote-name --location --insecure --silent --show-error "${FLUME_URL}" && \
+    tar zxf ${FLUME_TGZ} && \
+    mv apache-flume-${FLUME_VER}-bin ${FLUME_HOME} && \
+    mkdir -p ${FLUME_HOME}/plugins.d/cygnus && \
+    mkdir -p ${FLUME_HOME}/plugins.d/cygnus/lib && \
+    mkdir -p ${FLUME_HOME}/plugins.d/cygnus/libext && \
+    chown -R cygnus:cygnus ${FLUME_HOME}
+
+# Create Cygnus log folder
+RUN mkdir ${FLUME_HOME}/log
+
+# Clone fiware-cygnus
+RUN git clone ${GIT_URL_CYGNUS} ${CYGNUS_HOME} && \
+    cd ${CYGNUS_HOME} && \
+    git checkout ${GIT_REV_CYGNUS}
+
+# Build and install cygnus-common
+RUN cd ${CYGNUS_HOME}/cygnus-common && \
+    ${MVN_HOME}/bin/mvn clean compile exec:exec assembly:single && \
+    cp target/cygnus-common-${CYGNUS_VERSION}-jar-with-dependencies.jar ${FLUME_HOME}/plugins.d/cygnus/libext/ && \
+    ${MVN_HOME}/bin/mvn install:install-file -Dfile=${FLUME_HOME}/plugins.d/cygnus/libext/cygnus-common-${CYGNUS_VERSION}-jar-with-dependencies.jar -DgroupId=com.telefonica.iot -DartifactId=cygnus-common -Dversion=${CYGNUS_VERSION} -Dpackaging=jar -DgeneratePom=false
+
+# Build and install cygnus-ngsi
+RUN cd ${CYGNUS_HOME}/cygnus-ngsi && \
+    ${MVN_HOME}/bin/mvn clean compile exec:exec assembly:single && \
+    cp target/cygnus-ngsi-${CYGNUS_VERSION}-jar-with-dependencies.jar ${FLUME_HOME}/plugins.d/cygnus/lib/
+
+# Install Cygnus Application script
+RUN cp ${CYGNUS_HOME}/cygnus-common/target/classes/cygnus-flume-ng ${FLUME_HOME}/bin/ && \
+    chmod +x ${FLUME_HOME}/bin/cygnus-flume-ng
+
+# Instantiate some configuration files
+RUN cp ${CYGNUS_HOME}/cygnus-common/conf/log4j.properties.template ${FLUME_HOME}/conf/log4j.properties && \
+    cp ${CYGNUS_HOME}/cygnus-ngsi/conf/grouping_rules.conf.template ${FLUME_HOME}/conf/grouping_rules.conf
+
+# Cleanup to thin the final image
+RUN cd ${CYGNUS_HOME}/cygnus-common && \
```

Comment:
Docker builds the images by using a stack of layers.  Each of this layers is built from one of the directives on the Dockerfile (RUN, COPY, CMD, ...).  When you add a new layer, it records the changes needed to apply over the previous layer, but the previous layer is not modified at all, e.g:

```
# install program
RUN apt-get install program

# create myfile
RUN program --create-file myfile

# remove program
RUN apt-get remove program
```

This will create 3 layers:

* The first layer will contain the new files installed by the package.

* The second layer will contain just the changes that are the result of running the program (i.e. creates myfile).

* The third layer will mark the files from the package as removed, but it will not remove the files from the first layer, just hide them when using the image.

If you have several images that share some layers, this disk space will be shared, and that's good.  If you know that this won't happen, you can join this in a single RUN:

```
# install program
RUN apt-get install program && \

    # create myfile
    program --create-file myfile && \

    # remove program
    apt-get remove program
```

This will create a single layer with just the myfile file, as the package is removed in the same command that it is installed, thus not being added to the layer.

When using the image created with either of the two methods, the result will be the same, but the size of the image can be quite different, depending on the size of the package being installed.

In the case of cygnus, the removal of the build dependencies happens on a different layer than the one they are installed on, so the files are still present in the previous layer.  Though the files will not show up when using the image, they are still there and add to the size of the image.  Joining the RUN commands as shown in my second example will help in reducing the image size.